design-proposals/volume-update-strategy.md

```diff
@@ -0,0 +1,234 @@
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated, and the volumes need to be moved to another storage class.
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster.
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruptions in order to give it more storage.
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+This proposal introduces a new field in the VM spec to represent the volume update strategy called `updateVolumesStrategy`. This value is a unique field per VM, it can takes the values:
+ * `replacement` when the update of the volume aims to substitute the volume with another upon a VM restart
+ * `migration` when the update of the volume wants to migrate a volume to another one performing a full-copy of the content
+
+If no value is specified, the default behavior is `replacement`.
+
+Possibly, in the future, we could add a strategy for hotplugged volumes when the update wants to immediately add a new volume to a running VM.
+
+## API examples
+
+#### VM Spec API update
+
+Original VM spec
+```yaml
+  spec
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: src-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: src-pvc2
+  - name: vol3
+     persistentVolumeClaim:
+    claimName: src-pvc3
+```
+
+Update VM spec:
+```yaml
+spec:
+  updateVolumesStrategy: migration
+  volumes:
+  - name: vol1
+    persistentVolumeClaim:
+     claimName: dst-pvc1
+  - name: vol2
+    persistentVolumeClaim:
+     claimName: dst-pvc2
+  - name: vol3
+    persistentVolumeClaim:
+      claimName: src-pvc3
+```
+
+In this, example the user is able to update and migrate the first and second volume while the third one remains untouched.
+
+## Flow
+
+The volume migration update strategy follow the same [update sequence](https://github.com/kubevirt/community/blob/main/design-proposals/cpu-hotplug.md#phase-2-triggering-a-live-migration-when-the-number-of-sockets-changes-to-adjust-the-virt-launcher-cpu-resources) proposed for memory and cpu hotplug.
+
+Sequence of events if the `updateVolumesStrategy` is migration:
+
+* the user updates the volumes on a vm.spec.template.spec.domain.volumes
+* the VM controller determines the volumes which needs to be updated
+* the VM controller validates if the volumes can be migrated
+* the VM controller updates the volumes accordingly in the active vmi.Spec
+* the VMI controller writes the required volumes to migrate in the vmi.Status and set the update condition in the VMI status
```

Comment:
Thanks, just one last request: Please add an example where we have a few disks which now require an update. I simply wonder how many disks are mentioned in the condition, I do wonder if this needs to be parseable.

Reply:
@fabiand do you mean in the condition? Right now, I'm not mentioning any volumes in the condition, the information of the change is reported under migrated `migratedVolumes` and this is simply a list. So, it would look like:
```yaml
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2024-04-09T07:51:30Z"
      message: migrate volumes
      status: "True"
      type: VolumesChange
[..]
    migratedVolumes:
    - destinationPVCInfo:
        claimName: dest-pvc-1
        volumeMode: Block
      sourcePVCInfo:
        claimName: src-pvc-1
        volumeMode: Filesystem
    - destinationPVCInfo:
        claimName: dest-pvc-2
        volumeMode: Block
      sourcePVCInfo:
        claimName: src-pvc-2
        volumeMode: Filesystem
``` 

Do you want me to add this example or do you mean something else?

Reply:
Perfect. And yes, I was missing this example.

Witht his I'm good.

Reply:
Ok. I added the second volume in the example, so it is clearer